File: 42 Trapping Rain Water/42 Trapping Rain Water (Optimal).ts
function trapOptimzed(height: number[]): number {
  let total: number = 0;
  let leftMax: number = height[0];
  let rightMax: number = height[height.length - 1];
  let left: number = 0;
  let right: number = height.length - 1;

  while (left < right) {
    if (height[left] < height[right]) {
      leftMax = Math.max(leftMax, height[left]);
      total = total + Math.max(0, leftMax - height[left]);
      left++;
    } else {
      rightMax = Math.max(rightMax, height[right]);
      total = total + Math.max(0, rightMax - height[right]);
      right--;
    }
  }

  return total;
}

File: 42 Trapping Rain Water/42 Trapping Rain Water (Optimal).swift
class Solution {
  func trap(_ height: [Int]) -> Int {
    var total: Int = 0
    var leftMax: Int = height[0]
    var rightMax: Int = height.last!
    var left: Int = 0
    var right: Int = height.count - 1

    while left < right {
      if height[left] < height[right] {
        leftMax = max(leftMax, height[left])
        total += max(0, leftMax - height[left])
        left += 1
      } else {
        rightMax = max(rightMax, height[right])
        total += max(0, rightMax - height[right])
        right -= 1
      }
    }

    return total
  }
}